Front end

omzmarlon opened the pull request:
This pull request include:
 - Basic Authentication of our web app (required some minor backend change).
 - We can now use protobuf in typescript (see README.md, there's still some further work to make code generation more autonomous)
 (Our web app does not support any error handling yet. so there may have some unexpected behaviour)

omzmarlon committed:
Some re-structuring

omzmarlon committed:
Added some comments in AuthenticationController and fixed typo in UserService

omzmarlon committed:
Typescript protobuf setup

omzmarlon committed:
Changed method of generating typescript protobuf. The previous approach does not work

omzmarlon committed:
Updated README.md

omzmarlon committed:
Basic Authentication Completed. Now the web app can login, logout, register account.

Implemented a basic spinner.

More front/back end protobuf

Other infrastructure implemented.

tomyang729 commented:
辛苦了💦  我这两天花时间看一下

tomyang729 commented on play/src/main/scala/com/managementapp/services/common/ProtobufUtil.scala:
```diff
@@ -13,4 +13,9 @@ trait ProtobufUtil {
     builder.build.asInstanceOf[T]
   }
+
+  def generateJSON[M <: Message](content: M): String = {
```
The func name is a bit confusing the first time I saw it. Maybe pbToJson() is clearer? And the same for the one above. (jsonToPb() to keep them in consistent)

omzmarlon replied:
Agree

tomyang729 commented on web/src/app-container/authentication/services/authenticationService.ts:
```diff
@@ -0,0 +1,49 @@
+    constructor(private _http: MAHttp) {}
+
+    public isUserLoggedIn(): boolean {
+        return (localStorage.getItem(AuthenticationService.STORAGE_KEY) !== null)
```
Should we use cookie to store token? Localstoreage is generally not safe for sensitive data

omzmarlon replied:
True. And using cookies actually saves quite some code

omzmarlon committed:
Updated code based on review.
- Changed ProtobufUtil method naming
- switch to use cookies to implement authentication system

omzmarlon commented:
@tomyang729 Updated code based on review.
Renamed method in ProtobufUtil
And I switched to use cookies for authentication.
I didn't realize play has some nice support for cookie manipulation. Using cookies actually makes front-end and backend code cleaner.
And also don't need an "TokenDTO" anymore (we can just store username in cookies)

Reference: 
https://www.playframework.com/documentation/2.5.x/ScalaResults
http://alvinalexander.com/source-code/scala/how-getread-cookie-play-framework-2

tomyang729 commented:
👍 🏅
